### ISSUE TITLE
[Phase 2] Single Model Baseline with LightGBM

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/demand-forecasting-engine#3]
## Summary

Train a single LightGBM model to establish baseline performance for demand forecasting. This creates a reference point for evaluating ensemble improvements and validates the feature pipeline works end-to-end.

## System Context

```
demand-forecasting-engine/
├── src/
│   ├── models/
│   │   ├── __init__.py
│   │   ├── base_model.py       # <- CREATE
│   │   └── lightgbm_model.py   # <- CREATE
│   └── evaluation/
│       ├── __init__.py
│       └── metrics.py          # <- CREATE
└── scripts/
    └── train_baseline.py       # <- CREATE
```

## Dependencies

- Requires: Phase 0 (Data Foundation), Phase 1 (Feature Pipeline)
- Blocks: Phase 3 (Cross-Validation), Phase 4 (Quantile Regression)

### 1. Base Model Interface (`src/models/base_model.py`)

```python
from abc import ABC, abstractmethod
import numpy as np
import pandas as pd
from typing import Optional, Dict, Any


class BaseForecaster(ABC):
    """
    Abstract base class for all forecasting models.

    Provides consistent interface for training, prediction,
    and model inspection across different implementations.
    """

    @abstractmethod
    def fit(
        self,
        X: pd.DataFrame,
        y: pd.Series,
        eval_set: Optional[tuple] = None,
    ) -> 'BaseForecaster':
        """
        Train the model.

        Parameters
        ----------
        X : pd.DataFrame
            Feature matrix
        y : pd.Series
            Target values
        eval_set : tuple, optional
            (X_val, y_val) for early stopping

        Returns
        -------
        self
        """
        pass

    @abstractmethod
    def predict(self, X: pd.DataFrame) -> np.ndarray:
        """
        Generate predictions.

        Parameters
        ----------
        X : pd.DataFrame
            Feature matrix

        Returns
        -------
        np.ndarray
            Predicted values
        """
        pass

    @abstractmethod
    def get_feature_importance(self) -> pd.DataFrame:
        """
        Get feature importance scores.

        Returns
        -------
        pd.DataFrame
            Columns: ['feature', 'importance']
        """
        pass

    @abstractmethod
    def get_params(self) -> Dict[str, Any]:
        """Get model hyperparameters."""
        pass

    @abstractmethod
    def set_params(self, **params) -> 'BaseForecaster':
        """Set model hyperparameters."""
        pass
```

### 2. LightGBM Model (`src/models/lightgbm_model.py`)

```python
import numpy as np
import pandas as pd
import lightgbm as lgb
from typing import Optional, Dict, Any, List
from .base_model import BaseForecaster


class LightGBMForecaster(BaseForecaster):
    """
    LightGBM-based demand forecaster.

    Parameters
    ----------
    n_estimators : int
        Number of boosting rounds
    learning_rate : float
        Boosting learning rate
    max_depth : int
        Maximum tree depth (-1 for no limit)
    num_leaves : int
        Maximum number of leaves per tree
    min_child_samples : int
        Minimum samples required in a leaf
    feature_fraction : float
        Fraction of features to use per tree
    bagging_fraction : float
        Fraction of data to use per tree
    bagging_freq : int
        Frequency of bagging (0 = disable)
    reg_alpha : float
        L1 regularization
    reg_lambda : float
        L2 regularization
    random_state : int
        Random seed
    n_jobs : int
        Number of parallel jobs
    early_stopping_rounds : int
        Early stopping patience
    verbose : int
        Verbosity level
    """

    def __init__(
        self,
        n_estimators: int = 1000,
        learning_rate: float = 0.05,
        max_depth: int = -1,
        num_leaves: int = 31,
        min_child_samples: int = 20,
        feature_fraction: float = 0.8,
        bagging_fraction: float = 0.8,
        bagging_freq: int = 5,
        reg_alpha: float = 0.1,
        reg_lambda: float = 0.1,
        random_state: int = 42,
        n_jobs: int = -1,
        early_stopping_rounds: int = 50,
        verbose: int = -1,
    ):
        self.n_estimators = n_estimators
        self.learning_rate = learning_rate
        self.max_depth = max_depth
        self.num_leaves = num_leaves
        self.min_child_samples = min_child_samples
        self.feature_fraction = feature_fraction
        self.bagging_fraction = bagging_fraction
        self.bagging_freq = bagging_freq
        self.reg_alpha = reg_alpha
        self.reg_lambda = reg_lambda
        self.random_state = random_state
        self.n_jobs = n_jobs
        self.early_stopping_rounds = early_stopping_rounds
        self.verbose = verbose

        self.model_: Optional[lgb.Booster] = None
        self.feature_names_: Optional[List[str]] = None

    def fit(
        self,
        X: pd.DataFrame,
        y: pd.Series,
        eval_set: Optional[tuple] = None,
    ) -> 'LightGBMForecaster':
        """Train the LightGBM model."""
        self.feature_names_ = list(X.columns)

        train_data = lgb.Dataset(X, label=y)
        valid_sets = [train_data]

        if eval_set is not None:
            X_val, y_val = eval_set
            valid_data = lgb.Dataset(X_val, label=y_val, reference=train_data)
            valid_sets.append(valid_data)

        params = {
            'objective': 'regression',
            'metric': 'rmse',
            'boosting_type': 'gbdt',
            'learning_rate': self.learning_rate,
            'max_depth': self.max_depth,
            'num_leaves': self.num_leaves,
            'min_child_samples': self.min_child_samples,
            'feature_fraction': self.feature_fraction,
            'bagging_fraction': self.bagging_fraction,
            'bagging_freq': self.bagging_freq,
            'reg_alpha': self.reg_alpha,
            'reg_lambda': self.reg_lambda,
            'random_state': self.random_state,
            'n_jobs': self.n_jobs,
            'verbose': self.verbose,
        }

        callbacks = []
        if self.early_stopping_rounds and eval_set is not None:
            callbacks.append(lgb.early_stopping(self.early_stopping_rounds))

        self.model_ = lgb.train(
            params,
            train_data,
            num_boost_round=self.n_estimators,
            valid_sets=valid_sets,
            callbacks=callbacks,
        )

        return self

    def predict(self, X: pd.DataFrame) -> np.ndarray:
        """Generate predictions."""
        if self.model_ is None:
            raise ValueError("Model not fitted. Call fit() first.")
        return self.model_.predict(X)

    def get_feature_importance(self) -> pd.DataFrame:
        """Get feature importance scores."""
        if self.model_ is None:
            raise ValueError("Model not fitted. Call fit() first.")

        importance = self.model_.feature_importance(importance_type='gain')
        return pd.DataFrame({
            'feature': self.feature_names_,
            'importance': importance
        }).sort_values('importance', ascending=False)

    def get_params(self) -> Dict[str, Any]:
        """Get model hyperparameters."""
        return {
            'n_estimators': self.n_estimators,
            'learning_rate': self.learning_rate,
            'max_depth': self.max_depth,
            'num_leaves': self.num_leaves,
            'min_child_samples': self.min_child_samples,
            'feature_fraction': self.feature_fraction,
            'bagging_fraction': self.bagging_fraction,
            'bagging_freq': self.bagging_freq,
            'reg_alpha': self.reg_alpha,
            'reg_lambda': self.reg_lambda,
        }

    def set_params(self, **params) -> 'LightGBMForecaster':
        """Set model hyperparameters."""
        for key, value in params.items():
            if hasattr(self, key):
                setattr(self, key, value)
        return self
```

### 3. Evaluation Metrics (`src/evaluation/metrics.py`)

```python
import numpy as np
import pandas as pd
from typing import Dict


def mean_absolute_error(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    """Calculate Mean Absolute Error."""
    return np.mean(np.abs(y_true - y_pred))


def root_mean_squared_error(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    """Calculate Root Mean Squared Error."""
    return np.sqrt(np.mean((y_true - y_pred) ** 2))


def mean_absolute_percentage_error(
    y_true: np.ndarray,
    y_pred: np.ndarray,
    epsilon: float = 1e-8
) -> float:
    """
    Calculate Mean Absolute Percentage Error.

    Parameters
    ----------
    y_true : np.ndarray
        Actual values
    y_pred : np.ndarray
        Predicted values
    epsilon : float
        Small value to avoid division by zero

    Returns
    -------
    float
        MAPE as a decimal (0.1 = 10%)
    """
    return np.mean(np.abs((y_true - y_pred) / (np.abs(y_true) + epsilon)))


def symmetric_mape(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    """Calculate Symmetric Mean Absolute Percentage Error."""
    return np.mean(2 * np.abs(y_true - y_pred) / (np.abs(y_true) + np.abs(y_pred) + 1e-8))


def calculate_metrics(y_true: np.ndarray, y_pred: np.ndarray) -> Dict[str, float]:
    """
    Calculate all evaluation metrics.

    Returns
    -------
    dict
        Dictionary with metric names and values
    """
    return {
        'mae': mean_absolute_error(y_true, y_pred),
        'rmse': root_mean_squared_error(y_true, y_pred),
        'mape': mean_absolute_percentage_error(y_true, y_pred),
        'smape': symmetric_mape(y_true, y_pred),
    }


def print_metrics(metrics: Dict[str, float]) -> None:
    """Pretty print metrics."""
    print("\nEvaluation Metrics")
    print("=" * 30)
    print(f"MAE:   {metrics['mae']:.2f}")
    print(f"RMSE:  {metrics['rmse']:.2f}")
    print(f"MAPE:  {metrics['mape']:.2%}")
    print(f"SMAPE: {metrics['smape']:.2%}")
```

### 4. Temporal Split Utility (`src/evaluation/splits.py`)

```python
import pandas as pd
import numpy as np
from typing import Tuple


def temporal_train_test_split(
    X: pd.DataFrame,
    y: pd.Series,
    date_column: str = 'date',
    test_size: float = 0.2,
) -> Tuple[pd.DataFrame, pd.DataFrame, pd.Series, pd.Series]:
    """
    Split data temporally (no shuffling).

    Parameters
    ----------
    X : pd.DataFrame
        Feature matrix with date column
    y : pd.Series
        Target values
    date_column : str
        Name of date column
    test_size : float
        Fraction of data for test set

    Returns
    -------
    X_train, X_test, y_train, y_test
    """
    # Sort by date
    sorted_idx = X[date_column].argsort()
    X_sorted = X.iloc[sorted_idx].reset_index(drop=True)
    y_sorted = y.iloc[sorted_idx].reset_index(drop=True)

    # Calculate split point
    split_idx = int(len(X_sorted) * (1 - test_size))

    X_train = X_sorted.iloc[:split_idx]
    X_test = X_sorted.iloc[split_idx:]
    y_train = y_sorted.iloc[:split_idx]
    y_test = y_sorted.iloc[split_idx:]

    return X_train, X_test, y_train, y_test
```

### 5. Training Script (`scripts/train_baseline.py`)

```python
#!/usr/bin/env python
"""Train baseline LightGBM model."""
import argparse
import pandas as pd
import matplotlib.pyplot as plt

from src.data.loader import load_sales_data
from src.features.pipeline import build_feature_pipeline
from src.models.lightgbm_model import LightGBMForecaster
from src.evaluation.metrics import calculate_metrics, print_metrics
from src.evaluation.splits import temporal_train_test_split


def main():
    parser = argparse.ArgumentParser(description="Train baseline model")
    parser.add_argument("--data", required=True, help="Path to sales data")
    parser.add_argument("--output", default="models/baseline", help="Output directory")
    args = parser.parse_args()

    # Load data
    print("Loading data...")
    df = load_sales_data(args.data)

    # Build features
    print("Building features...")
    pipeline = build_feature_pipeline()
    df_features = pipeline.fit_transform(df)

    # Prepare X and y
    feature_cols = [c for c in df_features.columns if c not in ['date', 'sku_id', 'sales']]
    X = df_features[feature_cols].dropna()
    y = df_features.loc[X.index, 'sales']

    # Temporal split
    print("Splitting data...")
    X_train, X_test, y_train, y_test = temporal_train_test_split(
        X, y, test_size=0.2
    )
    print(f"Train size: {len(X_train):,}, Test size: {len(X_test):,}")

    # Train model
    print("Training LightGBM...")
    model = LightGBMForecaster()
    model.fit(X_train, y_train, eval_set=(X_test, y_test))

    # Evaluate
    print("\nEvaluating...")
    y_pred = model.predict(X_test)
    metrics = calculate_metrics(y_test.values, y_pred)
    print_metrics(metrics)

    # Feature importance
    importance = model.get_feature_importance()
    print("\nTop 10 Features:")
    print(importance.head(10).to_string(index=False))

    # Save results
    print(f"\nBaseline MAPE: {metrics['mape']:.2%}")


if __name__ == "__main__":
    main()
```

## Implementation Checklist

- [ ] Create `BaseForecaster` abstract class with consistent interface
- [ ] Implement `LightGBMForecaster` with training and prediction
- [ ] Add early stopping support with validation set
- [ ] Implement evaluation metrics (MAE, RMSE, MAPE, SMAPE)
- [ ] Create `temporal_train_test_split()` utility
- [ ] Create training script with CLI arguments
- [ ] Generate and save feature importance plot
- [ ] Document baseline metrics for future comparison

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/models/base_model.py` | 0-60 | Create | Abstract base class |
| `src/models/lightgbm_model.py` | 0-150 | Create | LightGBM implementation |
| `src/evaluation/metrics.py` | 0-70 | Create | Evaluation metrics |
| `src/evaluation/splits.py` | 0-50 | Create | Temporal split utility |
| `scripts/train_baseline.py` | 0-80 | Create | Training script |

## Technical Challenges

**Temporal Split**: Must not use random shuffling - future data cannot be in training set.

```python
# WRONG - shuffles data, leaks future
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2)

# CORRECT - respects time ordering
cutoff = X['date'].quantile(0.8)
X_train = X[X['date'] < cutoff]
X_test = X[X['date'] >= cutoff]
```

**Detecting Leakage**: If MAPE is unrealistically good (>95%), suspect leakage.

## Definition of Done

- [ ] LightGBM model trains without error
- [ ] Temporal split implemented correctly
- [ ] Metrics calculated and displayed
- [ ] Feature importance extractable
- [ ] Baseline MAPE documented
- [ ] Tests pass: `pytest tests/test_models.py -v`

## Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| Data leakage via random split | :red_circle: High | Use temporal_train_test_split only |
| Overfitting without validation | :yellow_circle: Medium | Use early stopping |
| Poor baseline indicating bad features | :yellow_circle: Medium | Review feature pipeline |

## Time Estimate

**Estimated Hours**: 8-10 hours
**Priority**: Critical (establishes performance baseline)
